### DISCUSSION TITLE
Using GraphSage to do unsupervised node embeddings for node regression

### DISCUSSION THREAD

**Discussion by lfangyu09** [pyg-team/pytorch_geometric#8654] [General]
Hi,

Thank you so much for this package. I want to use `GraphSage` to do unsupervised node embeddings for node regression. May I ask three questions:
1. The example [(examples/graph_sage_unsup.py)](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/graph_sage_unsup.py) seems to use for classification `loss = F.binary_cross_entropy_with_logits(pred, batch.edge_label)`. How can I change the loss function? Could it be changed to `loss = F.mse_loss(pred, batch.edge_label)`?
2. After training `GraphSage`, how can I use it to get node embedding? By `model(data.x, data.edge_index)`?
3.  Is `test()` necessary? As I use my own customed data, how to set `train_mask`, `val_mask`, and `test_mask`?

Thank you so much!

**Comment by rusty1s** [comment]
1. How do you want to do unsupervised learning for node regression? If you wanna use `mse_loss`, you need ground-truth targets.
2. Yes :)
3. No, it's not necessary. It's just implemented here to test the quality of node embeddings.

**Reply by lfangyu09** [reply]
Thanks a lot for your quick reply. 
1. I want to use `GraphSage` to get the positional encoding, similar to `AddRandomWalkPE` or `AddLaplacianEigenvectorPE`. Could I use this example [(examples/graph_sage_unsup.py)](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/graph_sage_unsup.py) to get the positional encoding?
2. May I ask what is the model output of this example [(examples/graph_sage_unsup.py)](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/graph_sage_unsup.py)? 

Thanks.

**Reply by rusty1s** [reply]
Yes, you can basically use `graph_sage_unsup.py` as it is to get pre-trained node embeddings, which you can then use to perform a final regression downstream task. The model output here are basically pre-trained node embeddings.
